apps/calendar/js/utils/overlap.js

```diff
@@ -1,135 +1,167 @@
-      } else if (myCS && otherCS && myCS != otherCS) {
-        // Both already in different spans, so absorb other into this
-        myCS.absorb(otherCS);
-      }
-    });
-
+  /**
+   * add busytime and element to the overlap calculation
+   * @param {object} item Object containing {busytime} and {element} properties
+   */
+  add: function(item) {
+    this._items.push(item);
   },
 
   /**
-   * Remove a busytime from the collection.
-   * Unlike other methods you must pass a real
-   * busytime object.
-   *
-   * @param {Object} busytime full busytime object.
+   * update the element width, position and classList based on overlaps
    */
-  remove: function(busytime) {
-    this._clearLayout(busytime);
-    this.tree.remove(busytime);
-    delete this.elements[busytime._id];
-    var myID = busytime._id;
-    var myCS = this.conflicts[myID];
-    if (myCS) {
-      myCS.remove(busytime);
-    }
+  render: function() {
+    // we defer the conflict work till render to avoid calculating it multiple
+    // times in a row (eg. multiple `add` calls)
+    this._updateConflicts();
+    this._getConflictItems().forEach(data => {
+      var el = data.element;
+      var nCols = data.nCols;
+      var wid = (100 / nCols);
+      el.style.width = wid + '%';
+      // on LTR we show earlier events on the left, on RTL we do the opposite
+      var position = document.documentElement.dir === 'rtl' ? 'right' : 'left';
+      el.style[position] = (data.colIndex * wid) + '%';
+      // we toggle the style based on amount of overlaps
+      el.classList.toggle('many-overlaps', nCols > 4);
+      el.classList.toggle('has-overlaps', nCols > 1);
+    });
   },
 
   /**
-   * Get the ConflictSpan associated with this busytime, if any.
-   *
-   * @param {Object|String} busytime id or busytime object.
-   * @return {Object} associated ConflictSpan, if any.
+   * (re)builds all the ConflictSpan
    */
-  getConflictSpan: function(busytime) {
-    var id = this._busytimeId(busytime);
-    return this.conflicts[id];
+  _updateConflicts: function() {
+    // need to reset the conflicts since new items might be added between
+    // multiple render calls
+    this._conflicts = [];
+    // sort by start date to avoid issues with larger events being added
+    // afterwards and also to ensure events that started earlier are on the
+    // left side on LTR and right side on RTL (if possible)
+    this._items
+      .sort((a, b) => a.busytime.startDate - b.busytime.startDate)
+      .forEach(item => this._getConflict(item).add(item));
   },
 
   /**
-   * @param {Object|String} busytime id or busytime object.
-   * @return {HTMLElement} associated dom element.
+   * gets ConflictSpan that can fit the item or create a new ConflictSpan
    */
-  getElement: function(busytime) {
-    var id = this._busytimeId(busytime);
-    return this.elements[id];
+  _getConflict: function(item) {
+    var conf, i = -1;
+    while ((conf = this._conflicts[++i])) {
```

Comment:
because jshint doesn't allow assignments inside `while()` unless you set `boss:true` to avoid errors like `while (foo = bar)` when you really mean `while (foo == bar)`.